File: 02_DB_Interpolation/smallwing/matlab_code/matlab_func/load_int_file.m
function [ bb_buff ] = load_int_file( path_in, variablename, i_file, history, ncores )
%UNTITLED2 Summary of this function goes here
%   Detailed explanation goes here

fname         = strcat(path_in,variablename,num2str(i_file,'%12.5d'));
[fid,~] = fopen(fname,'r','ieee-le');
hdr           = fread(fid,1,'int32')      ;
F             = fread(fid,hdr,'*char')'   ;
dum5          = fread(fid,1,'*float64')   ;
Rer           = fread(fid,1,'*float64')   ;
Domain        = fread(fid,3,'*float64')   ;
nel           = fread(fid,3,'int32')      ;
Poly          = fread(fid,3,'int32')      ;
% nstat         = fread(fid,1,'int32')      ;
% nderiv        = fread(fid,1,'int32')      ;
time         = fread(fid,1,'*float64')   ;
% timee         = fread(fid,1,'*float64')   ;
% atime         = fread(fid,1,'*float64')   ;
% DT            = fread(fid,1,'*float64')   ;
% nrec          = fread(fid,1,'int32')      ;
% Tint          = fread(fid,1,'*float64')   ;
npoints       = fread(fid,1,'int32')      ;
nfields       = fread(fid,1,'int32')      ;
dum6          = fread(fid,1,'*float64')   ;

disp(fname)
aa=fread(fid,Inf,'*float64'); 
%Assign points taking into account gaps among processors

disp(['read ',fname])
for i=1:ncores
    bb_buff(1+sum(history(1:i-1,2)):sum(history(1:i,2)),1)=aa(1+sum(history(1:i-1,2))+(i-1):sum(history(1:i,2))+(i-1));
    test_read=bb_buff(1+sum(history(1:i-1,2)):sum(history(1:i,2)));
    minrankval=min(test_read);
    maxrankval=max(test_read);
    disp(['irank=',num2str(i-1),' min=',num2str(minrankval),' max=',num2str(max(maxrankval))])
    disp(test_read(1:10))
    disp(test_read(end-10:end))
end


end

File: DB_Interpolation/NACA4412_5deg/3_convert_to_hdf5/load_int_file.py
import numpy as np
import scipy.io as sio

def load_int_file(path_in: str, variable_name: str, 
                  i_file: int, history: np.ndarray, ncores: int) -> np.ndarray[float]:
    """
    Reads a binary data file and extracts relevant information.
    
    Parameters:
    - path_in (str): Input file path.
    - variable_name (str): Variable name prefix.
    - i_file (int): File index.
    - history (numpy array): Array containing history of data distribution.
    - ncores (int): Number of processing cores.
    
    Returns:
    - bb_buff (numpy array): Extracted data.
    """
    fname = path_in + "/" + variable_name + f"{i_file:05d}"
    with open(fname, "rb") as fid:
        # print(f"  - Reading {fname}...", flush=True)

        # Read header
        hdr     = np.fromfile(fid, dtype=np.int32,   count=1)[0]
        F       = np.fromfile(fid, dtype=np.byte,    count=hdr)
        dum5    = np.fromfile(fid, dtype=np.float64, count=1)[0]
        Rer     = np.fromfile(fid, dtype=np.float64, count=1)[0]
        Domain  = np.fromfile(fid, dtype=np.float64, count=3)
        nel     = np.fromfile(fid, dtype=np.int32,   count=3)
        Poly    = np.fromfile(fid, dtype=np.int32,   count=3)
        time    = np.fromfile(fid, dtype=np.float64, count=1)[0]
        npoints = np.fromfile(fid, dtype=np.int32,   count=1)[0]
        nfields = np.fromfile(fid, dtype=np.int32,   count=1)[0]
        dum6    = np.fromfile(fid, dtype=np.float64, count=1)[0]

        # Read remaining data
        aa = np.fromfile(fid, dtype=np.float64)

        bb_buff = np.zeros(np.sum(history[:, 1]), dtype=np.float64)
        for i in range(ncores):
            start_ibuff = np.sum(history[:i, 1])
            end_ibuff   = np.sum(history[:i+1, 1])
            # print(f"{start_ibuff = }")
            # print(f"{end_ibuff = }")
            start_iaa   = start_ibuff + i
            end_iaa     = end_ibuff + i
            # print(f"{start_iaa = }")
            # print(f"{end_iaa = }")

            # Adjusting for processor gaps within aa
            bb_buff[start_ibuff:end_ibuff] = aa[start_iaa:end_iaa]

            # # Debugging
            # test_read = bb_buff[start_ibuff:end_ibuff]
            # minrankval = np.min(test_read)
            # maxrankval = np.max(test_read)
            # print(f"irank={i}, min={minrankval}, max={maxrankval}")
            # print(test_read[:10])
            # print(test_read[-10:])

    return bb_buff, time
